design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,380 @@
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider
+   ...
+      env:
+        - name: K8S_SECRET_LABELS
+          value: env-prod,env-test
+```
+
+These labels will have an OR relationship not an AND relationship. In other words, a K8s Secret can either have a CONJUR label or PROD label. A customer can also write '*' to imply all K8s secrets in the namespace. If 'K8S_SECRET_LABELS' is not supplied, by default, the Secrets Provider will fetch all K8s Secrets in the namespace.
+
+Decision: **Solution #1, Read all K8s Secrets that are marked with labels**
+
+We decided against listing K8s Secrets as configuration in the ENV of the Secrets Provider manifest because that is not a scalable solution, especially for customers with hundreds of K8s Secrets. Customers might already use labels for grouping related K8s Secrets so we will benefit by integrating more with a customer environment and utilizing that.
+
+
+
+### Milestone 2: Update K8s secret with Conjur secret
+
+#### Design
+
+Once a secret has been updated in Conjur, this will trigger an update in K8s Secrets values. For Milestone 2, we are building ontop of Milestone 1, **not** in-place of.  The Secrets Provider deployment will need to be a Pod that does not terminate (instead of a Job). Every change in Conjur secret, will trigger an event to update the K8s Secret. 
+
+Note that we are not introducing a breaking change between Milestones. If a customer does not want to upgrade to Milestone 2, they are not required to. If they do, the image they will use will be the same one used in Milestone 1.
+
+|      | Trigger for Secret Rotation                                  | Elaboration                                                  | Pros                                                         | Cons                                                         |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | *Every 5 minutes,* retrieve secrets every constant, configurable amount of time. Possible options:<br /><br />a. Run constantly and wake up when time is right <br />b. Spin-up every 5 minutes, run using [cron job](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) based on K8s secrets update (Get trigger upon secret change) | *Flow*<br />1. Every 5 minutes, Secrets Provider checks if a change/a new K8s Secret was created <br />2. If so, update Secrets Provider local cache of state for K8s | - Easily integrate with current solution since retry mechanism already exists<br />- No need to change anything in Conjur | -  Polling is a bad way for changes notification mechanism.<br />- Option (a) is inefficient management of resources.<br />- Option (b) is supported only since K8s version 1.18 and is currently in beta. |
```

Comment:
how does the `Secrets Provider check if a change/a new K8s Secret was created`? do we save some state?

or does it always retrieve all k8s secrets and perform the whole cycle?

Reply:
Removed this for now since we will focus on Milestone 1